Print react improvements

carlobeltrame opened the pull request:
Resolves the following points from #1241:

### Tech
- Can we do better than re-implement small read-only versions of hal-json-vuex and vue-i18n?
  - [x] If we upgrade to [vue-i18n v9](https://vue-i18n.intlify.dev/), I think this will be possible: It has much better separation of concerns, and we should be able to instantiate vue-i18n without adding it to a Vue instance. ~~But vue-i18n v9 only supports Vue 3...~~ By using only the core library, Vue is not needed.
- [x] Fix the console errors about requestAnimationFrame when using the web worker in Firefox
- [x] Can't import named exports in jsx files:
  ```js
   // does not work
  import { sortBy } from 'lodash'
  // works
  import lodash from 'lodash'
  const { sortBy } = lodash
  ```
- [x] Sentry in web worker
- [x] Contribute globalWorkerShim.js to exogee-technology/vite-plugin-shim-react-pdf https://github.com/exogee-technology/vite-plugin-shim-react-pdf/pull/3

### Features
- [x] Better "progress" indicators
- [x] Display rendering errors in LocalPdfDownloadButton and LocalPrintPreview to the user
- [x] Respect the print config (checkboxes) and adapt data fetching and output filename depending on the print config

carlobeltrame committed:
Update shim and remove our own extension

Now that the content of our shim extension was merged in
https://github.com/exogee-technology/vite-plugin-shim-react-pdf/pull/3
we can safely remove our own shimming.

carlobeltrame committed:
Automatically re-render the preview when the locale changes

carlobeltrame committed:
Use the core of vue-i18n v9 to translate messages in worker without requiring Vue

carlobeltrame committed:
Refactor to only import React when necessary

React and other rendering dependencies are now only imported when
absolutely necessary. If rendering the PDF in the main thread, this
means React is imported via a dynamic import. If rendering in a web
worker, React is imported statically in the worker script, but the
worker script itself is imported dynamically.

Also introduces a clearer directory structure. "documents" are like
"views" in HTML rendering, i.e. root components which may contain other
components from the "components" directory.

carlobeltrame committed:
Add feedback for users on print error

carlobeltrame committed:
Prevent multiple concurrent preview workers

carlobeltrame committed:
Make it more obvious how to switch between worker and main thread in development

carlobeltrame committed:
Update url-template to 3.0.0

carlobeltrame committed:
Simplify imports

Works around missing named exports in react-pdf, see
https://github.com/diegomura/react-pdf/issues/1317

carlobeltrame committed:
Fix React warnings for missing requestAnimationFrame in Firefox workers

Unfortunately, this requires us to bring back globalWorkerShim.js,
because the code in vite-plugin-shim-react-pdf only runs immediately
before React, and not before importing the raf polyfill.

carlobeltrame committed:
Silence more unnecessary warnings

carlobeltrame committed:
Render different sections depending on the print config

carlobeltrame committed:
Let the print document component determine the file name

carlobeltrame committed:
Work around ESLint limitation

carlobeltrame reviewed (commented):
Auf dem Deployment bekommt man in Firefox noch folgenden Konsolenfehler beim PDF Generieren:
```
DataCloneError: The object could not be cloned.
```
Muss ich mir noch ansehen warum das passiert.

pmattmann commented:
Auf https://pr2478.ecamp3.ch/ bekomme ich zudem folgender Fehler:
```
Error: Unknown font format
    at Object.fontkit$1.create (renderPdf.worker.3efec10f.js:56:127226)
    at s.<anonymous> (renderPdf.worker.3efec10f.js:85:5836)
    at ce (renderPdf.worker.3efec10f.js:42:19583)
    at Generator._invoke (renderPdf.worker.3efec10f.js:42:21758)
    at Generator.next (renderPdf.worker.3efec10f.js:42:20126)
    at asyncGeneratorStep (renderPdf.worker.3efec10f.js:42:18554)
    at b (renderPdf.worker.3efec10f.js:42:18769)
```

Habe es sowohl mit Chrome wie auch mit Edge versucht.

carlobeltrame committed:
Add sentry to worker and fix firefox cloning issue

If an error happened in the web worker, this error would be serialized
as an error object into the result returned from the web worker. Firefox
did not like this, while the other browsers were silent about it. But
really, only plain JSON serializable data may be passed between worker
and main thread (although blobs seem to be okay).
So instead, we use Sentry to log any error details in the worker, and
pass only the message string back to the main thread, to be logged or
displayed to the user.

carlobeltrame committed:
Fix 'Unknown font format' error in production

Vite would previously simply omit the font files when bundling for
production. When react-pdf went on to fetch the fonts, vite would then
serve the index.html file instead of the font files. This led to the
Unknown font format error.
The simple fix in this commit forces vite to bundle the font files
during npm run build.

carlobeltrame committed:
Fix import error in firefox during development again

'Syntax Error: import declarations may only appear at top level of a
module' is otherwise thrown in firefox during development.
This is the reason why we need the vite-plugin-worker dependency. This
vite plugin polyfills missing logic in vite for making our workers work
across a wider range of browsers.

carlobeltrame committed:
Fix frontend build after rebase

carlobeltrame commented:
> Auf https://pr2478.ecamp3.ch/ bekomme ich zudem folgender Fehler:
> 
> ```
> Error: Unknown font format
>     at Object.fontkit$1.create (renderPdf.worker.3efec10f.js:56:127226)
>     at s.<anonymous> (renderPdf.worker.3efec10f.js:85:5836)
>     at ce (renderPdf.worker.3efec10f.js:42:19583)
>     at Generator._invoke (renderPdf.worker.3efec10f.js:42:21758)
>     at Generator.next (renderPdf.worker.3efec10f.js:42:20126)
>     at asyncGeneratorStep (renderPdf.worker.3efec10f.js:42:18554)
>     at b (renderPdf.worker.3efec10f.js:42:18769)
> ```
> 
> Habe es sowohl mit Chrome wie auch mit Edge versucht.

Fixed in c5816eb128ec589351655207e4f661e97855b561. Problem lag bei vite. Die Font-Dateien wurden beim Prod-Build nicht gebundelt, weil vite wies aussieht Asset-Dependencies von .jsx Dateien nicht richtig resolved (?). Lösung war, die benötigten Font-Dateien auch in einer normalen .js-Datei mal zu importieren.

Getestet in Firefox und Chromium auf Ubuntu, und Firefox und Edge auf Windows.

> Auf dem Deployment bekommt man in Firefox noch folgenden Konsolenfehler beim PDF Generieren:
> 
> ```
> DataCloneError: The object could not be cloned.
> ```
> 
> Muss ich mir noch ansehen warum das passiert.

Gefixt in a36da2c9aabbd7fd4673979ff35d0a9c0fa37095. Problem war, dass ich im Fall eines Fehlers beim Rendern bisher versucht hatte, das gecatchte `Error`-Objekt vom Worker zurück an den Main Thread zu geben. Firefox hat sich (korrekterweise) beklagt, dass dieses Objekt nicht für die Übertragung zwischen Threads geeignet ist, und den obigen Fehler geloggt. Die anderen Browser haben soweit ich beurteilen kann den Fehler einfach "verschluckt". Lösung war, den Fehler stattdessen direkt im Worker via Sentry zu loggen, und nur das `error.message` String an den Main Thread zurückzugeben.
